File: 09_dynamic_programming/python/02_longest_common_substring.py
dp_table_blue = ["b", "l", "u", "e"]
dp_table_clues = ["c", "l", "u", "e", "s"]
dp_table = [[0 for i in range(len(dp_table_blue))] for i in range(len(dp_table_clues))]
print(dp_table)  

for i in range(0, len(dp_table_clues)):
    for j in range(0, len(dp_table_blue)):
        if dp_table_clues[i] == dp_table_blue[j]:
            dp_table[i][j] = dp_table[i - 1][j - 1] + 1
        else:
            dp_table[i][j] = 0

print(dp_table)

File: 11_dynamic_programming/scala/02_longest_common_substring.scala
val dpTableBlue = Array("b", "l", "u", "e")
val dpTableClues = Array("c", "l", "u", "e", "s")

// Create a 2D array initialized with zeros
val dpTable = Array.ofDim[Int](dpTableBlue.length, dpTableClues.length)

// for each row
for(i <- 0 until dpTableBlue.length) {
  // for each column
  for(j <- 0 until dpTableClues.length) {
    if(dpTableClues(j) == dpTableBlue(i)) {
      val prevVal = if(i > 0 && j > 0) dpTable(i-1)(j-1) else 0
      dpTable(i)(j) = prevVal + 1
    } else {
      dpTable(i)(j) = 0
    }
  }
}

// Display table
for(row <- dpTable) {
  println(row.mkString(" "))
}

